From 489532d428f655bed8853fd70eb97438224573c5 Mon Sep 17 00:00:00 2001
From: Daniel Leung <daniel.leung@intel.com>
Date: Thu, 6 Dec 2012 13:48:18 -0800
Subject: x86: Enable building x86_64 kernel

Enable building the x86_64 kernel using the x86_64 host toolchain
(for now, until there is an official x86_64 target toolchain).
This allows us to have 64-bit kernel, and still have 32-bit
Android userspace.

For: AXIA-1343
Change-Id: Icc76966fc311bc170c04653d5356eaec11de4f52
Signed-off-by: Daniel Leung <daniel.leung@intel.com>
---
 core/combo/TARGET_linux-x86.mk |    8 ++++++++
 core/dumpvar.mk                |    1 +
 core/kernel.mk                 |    6 +++---
 3 files changed, 12 insertions(+), 3 deletions(-)

diff --git a/core/combo/TARGET_linux-x86.mk b/core/combo/TARGET_linux-x86.mk
index 712cfc6..80d3095 100644
--- a/core/combo/TARGET_linux-x86.mk
+++ b/core/combo/TARGET_linux-x86.mk
@@ -46,6 +46,14 @@ TARGET_TOOLCHAIN_ROOT := prebuilts/gcc/$(HOST_PREBUILT_TAG)/x86/i686-linux-andro
 TARGET_TOOLS_PREFIX := $(TARGET_TOOLCHAIN_ROOT)/bin/i686-linux-android-
 endif
 
+ifeq ($(TARGET_KERNEL_ARCH),x86_64)
+TARGET_KERNEL_TOOLCHAIN_ROOT := prebuilts/gcc/$(HOST_PREBUILT_TAG)/host/x86_64-linux-glibc2.7-4.6
+TARGET_KERNEL_TOOLS_PREFIX := $(TARGET_KERNEL_TOOLCHAIN_ROOT)/bin/x86_64-linux-
+else
+TARGET_KERNEL_TOOLCHAIN_ROOT := $(TARGET_TOOLCHAIN_ROOT)
+TARGET_KERNEL_TOOLS_PREFIX := $(TARGET_TOOLS_PREFIX)
+endif
+
 TARGET_CC := $(TARGET_TOOLS_PREFIX)gcc$(HOST_EXECUTABLE_SUFFIX)
 TARGET_CXX := $(TARGET_TOOLS_PREFIX)g++$(HOST_EXECUTABLE_SUFFIX)
 TARGET_AR := $(TARGET_TOOLS_PREFIX)ar$(HOST_EXECUTABLE_SUFFIX)
diff --git a/core/dumpvar.mk b/core/dumpvar.mk
index 92bf6af..ef601e6 100644
--- a/core/dumpvar.mk
+++ b/core/dumpvar.mk
@@ -99,6 +99,7 @@ $(info   TARGET_BUILD_APPS=$(TARGET_BUILD_APPS))
 $(info   TARGET_ARCH=$(TARGET_ARCH))
 $(info   TARGET_ARCH_VARIANT=$(TARGET_ARCH_VARIANT))
 $(info   TARGET_CPU_VARIANT=$(TARGET_CPU_VARIANT))
+$(info   TARGET_KERNEL_ARCH=$(TARGET_KERNEL_ARCH))
 $(info   HOST_ARCH=$(HOST_ARCH))
 $(info   HOST_OS=$(HOST_OS))
 $(info   HOST_OS_EXTRA=$(HOST_OS_EXTRA))
diff --git a/core/kernel.mk b/core/kernel.mk
index 39cd0a0..34ef3e0 100644
--- a/core/kernel.mk
+++ b/core/kernel.mk
@@ -94,11 +94,11 @@ modbuild_output := $(CURDIR)/$(TARGET_OUT_INTERMEDIATES)/kernelmods
 # are out of alignment.
 mk_kernel := + $(hide) PATH=$(CURDIR)/build/tools/gzip_hack/:$(PATH) $(MAKE) -C $(TARGET_KERNEL_SOURCE)  O=$(kbuild_output) ARCH=$(TARGET_KERNEL_ARCH) $(if $(SHOW_COMMANDS),V=1) KCFLAGS="$(TARGET_KERNEL_EXTRA_CFLAGS)"
 ifneq ($(TARGET_KERNEL_CROSS_COMPILE),false)
-ifneq ($(TARGET_TOOLS_PREFIX),)
+ifneq ($(TARGET_KERNEL_TOOLS_PREFIX),)
 ifneq ($(USE_CCACHE),)
-mk_kernel += CROSS_COMPILE="$(CCACHE_BIN) $(CURDIR)/$(TARGET_TOOLS_PREFIX)"
+mk_kernel += CROSS_COMPILE="$(CCACHE_BIN) $(CURDIR)/$(TARGET_KERNEL_TOOLS_PREFIX)"
 else
-mk_kernel += CROSS_COMPILE=$(CURDIR)/$(TARGET_TOOLS_PREFIX)
+mk_kernel += CROSS_COMPILE=$(CURDIR)/$(TARGET_KERNEL_TOOLS_PREFIX)
 endif
 endif
 endif
-- 
1.7.4.1

